File: src/scanner.py
"""
Directory scanning functionality for the Directory Analyzer.

This module handles the core logic for traversing directories and calculating sizes.
"""

__all__ = ["DirectoryScanner", "estimate_directory_count"]

import logging
from collections.abc import Iterator
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime
from pathlib import Path

from .classifier import ContentClassifier
from .models import DirectoryInfo, EnhancedDirectoryInfo, ScanOptions, ScanStatistics
from .utils import (
    ProgressReporter,
    get_direct_files,
    get_subdirectories,
    is_accessible_directory,
    safe_get_file_size,
    write_error_log,
)


class DirectoryScanner:
    """Handles scanning directories and calculating sizes with full file classification."""

    def __init__(self, options: ScanOptions):
        """Initialize the scanner with configuration options.

        Args:
            options: Scanning configuration options
        """
        self.options = options
        self.logger = logging.getLogger(__name__)
        self.error_directories = []  # Track inaccessible directories
        self.classifier = ContentClassifier()  # File classification system

        # Statistics tracking
        self.total_directories = 0
        self.total_files = 0
        self.total_size = 0
        self.all_files = []  # Store all file information for global statistics

    def scan_single_directory(self, directory: Path) -> DirectoryInfo:
        """Scan a single directory and calculate its direct file size with classification.

        Args:
            directory: Directory path to scan

        Returns:
            DirectoryInfo with calculated size and file count
        """
        scan_time = datetime.now()

        try:
            if not is_accessible_directory(directory):
                error_msg = "Permission denied or directory inaccessible"
                self.error_directories.append(str(directory))
                return DirectoryInfo(
                    path=directory,
                    size_bytes=0,
                    file_count=0,
                    last_scanned=scan_time,
                    error_message=error_msg
                )

            total_size = 0
            file_count = 0
            directory_files = []
              # Process each file with classification
            for file_path in get_direct_files(directory, self.options.verbose):# Apply extension filter if specified
                if self.options.extension_filter:
                    if file_path.suffix.lower() not in self.options.extension_filter:
                        continue

                file_size = safe_get_file_size(file_path)
                total_size += file_size
                file_count += 1

                # Create file info with classification
                file_info = self.classifier.classify_file_with_info(file_path, file_size)
                directory_files.append(file_info)
                self.all_files.append(file_info)
                  # Update global statistics
                self.total_files += 1
                self.total_size += file_size

            return DirectoryInfo(
                path=directory,
                size_bytes=total_size,
                file_count=file_count,
                last_scanned=scan_time
            )

        except Exception as e:
            # Only show individual warnings in verbose mode, but always track for summary
            if self.options.verbose:
                self.logger.warning(f"Error scanning directory {directory}: {e}")
            else:
                self.logger.debug(f"Error scanning directory {directory}: {e}")

            self.error_directories.append(str(directory))
            return DirectoryInfo(
                path=directory,
                size_bytes=0,
                file_count=0,
                last_scanned=scan_time,
                error_message=str(e)
            )

    def get_all_directories(self, root_path: Path) -> Iterator[Path]:
        """Recursively get all directories under the root path.

        Args:
            root_path: Root directory to start scanning from

        Yields:
            Path objects for all directories found
        """
        # Include the root directory itself
        self.total_directories += 1
        yield root_path
          # Recursively find all subdirectories
        try:
            for subdirectory in get_subdirectories(root_path, self.options.include_hidden, self.options.verbose):
                yield from self.get_all_directories(subdirectory)
        except Exception as e:
            # Only show traversal warnings in verbose mode
            if self.options.verbose:
                self.logger.warning(f"Error traversing directory {root_path}: {e}")
            else:
                self.logger.debug(f"Error traversing directory {root_path}: {e}")
            self.error_directories.append(str(root_path))

    def scan_directories_sequential(self) -> list[DirectoryInfo]:
        """Scan all directories sequentially.

        Returns:
            List of DirectoryInfo objects
        """
        results = []

        if self.options.verbose:
            print(f"Starting sequential scan of {self.options.target_path}")

        # Get all directories first to show progress
        all_directories = list(self.get_all_directories(self.options.target_path))
        total_dirs = len(all_directories)

        progress = None
        if self.options.verbose:
            progress = ProgressReporter(total_dirs, "Scanning directories")

        for directory in all_directories:
            result = self.scan_single_directory(directory)

            # Apply size filter
            if result.size_bytes >= self.options.min_size_bytes:
                results.append(result)

            if progress:
                progress.update()

        if progress:
            progress.finish()
            print(f"Scanned {total_dirs} directories, {len(results)} meet size criteria")

        return results

    def scan_directories_parallel(self, max_workers: int = 4) -> list[DirectoryInfo]:
        """Scan directories in parallel using thread pool.

        Args:
            max_workers: Maximum number of worker threads

        Returns:
            List of DirectoryInfo objects
        """
        results = []

        if self.options.verbose:
            print(f"Starting parallel scan of {self.options.target_path} with {max_workers} workers")

        # Get all directories first
        all_directories = list(self.get_all_directories(self.options.target_path))
        total_dirs = len(all_directories)

        progress = None
        if self.options.verbose:
            progress = ProgressReporter(total_dirs, "Scanning directories")

        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            # Submit all scanning tasks
            future_to_directory = {
                executor.submit(self.scan_single_directory, directory): directory
                for directory in all_directories
            }

            # Process results as they complete
            for future in as_completed(future_to_directory):
                try:
                    result = future.result()

                    # Apply size filter
                    if result.size_bytes >= self.options.min_size_bytes:
                        results.append(result)

                    if progress:
                        progress.update()

                except Exception as e:
                    directory = future_to_directory[future]
                    self.logger.exception(f"Error processing directory {directory}: {e}")

                    if progress:
                        progress.update()

        if progress:
            progress.finish()
            print(f"Scanned {total_dirs} directories, {len(results)} meet size criteria")

        return results

    def scan(self, use_parallel: bool = True, max_workers: int = 4) -> list[DirectoryInfo]:
        """Scan directories and return results.

        Args:
            use_parallel: Whether to use parallel processing
            max_workers: Maximum number of worker threads for parallel processing

        Returns:
            List of DirectoryInfo objects sorted by size (largest first)
        """
        self.logger.info(f"Starting directory scan of {self.options.target_path}")

        try:
            if use_parallel and max_workers > 1:
                results = self.scan_directories_parallel(max_workers)
            else:
                results = self.scan_directories_sequential()

            # Sort by size (largest first)
            results.sort(key=lambda x: x.size_bytes, reverse=True)

            # Write error log if there were any inaccessible directories
            if self.error_directories:
                write_error_log(self.options.error_log_file, self.error_directories)

            self.logger.info(f"Scan completed. Found {len(results)} directories.")
            return results

        except KeyboardInterrupt:
            self.logger.info("Scan interrupted by user")
            raise
        except Exception as e:
            self.logger.exception(f"Scan failed: {e}")
            raise

    def create_scan_statistics(self, results: list[DirectoryInfo], scan_duration: float) -> ScanStatistics:
        """Create comprehensive scan statistics with content classification.

        Args:
            results: List of directory scan results
            scan_duration: Time taken for the scan

        Returns:
            ScanStatistics object with comprehensive data including category breakdown
        """
        # Calculate category breakdown from all files
        category_breakdown = self.classifier.get_category_statistics(self.all_files)
        file_count_by_category = self.classifier.get_file_count_by_category(self.all_files)

        return ScanStatistics(
            total_directories=self.total_directories,
            total_files=self.total_files,
            total_size_bytes=self.total_size,
            scan_duration=scan_duration,
            category_breakdown=category_breakdown,
            file_count_by_category=file_count_by_category
        )

    def create_enhanced_directory_info(self, directory: DirectoryInfo, total_size: int) -> EnhancedDirectoryInfo:
        """Create enhanced directory information with classification and percentages.

        Args:
            directory: Base directory information
            total_size: Total size of all scanned content for percentage calculation

        Returns:
            EnhancedDirectoryInfo with file classification and percentage analysis
        """
        # Find files that belong to this directory
        directory_files = [f for f in self.all_files if f.path.parent == directory.path]

        # Calculate category breakdown for this directory
        category_breakdown = self.classifier.get_category_statistics(directory_files)

        # Determine dominant category
        dominant_category = self.classifier.get_dominant_category(directory_files) if directory_files else "other"

        # Calculate percentage of total
        percentage_of_total = (directory.size_bytes / total_size * 100) if total_size > 0 else 0.0

        return EnhancedDirectoryInfo(
            base_info=directory,
            files=directory_files,
            category_breakdown=category_breakdown,
            dominant_category=dominant_category,
            percentage_of_total=percentage_of_total
        )


def estimate_directory_count(path: Path, max_depth: int = 3) -> int:
    """Estimate the total number of directories under a path.

    This is used for progress reporting and resource planning.

    Args:
        path: Root path to estimate
        max_depth: Maximum depth to sample for estimation

    Returns:
        Estimated number of directories
    """
    try:
        def count_directories(current_path: Path, depth: int) -> int:
            if depth > max_depth:
                # Estimate based on current sample
                if depth == 1:
                    return 1
                # Use exponential estimation
                return int(1.5 ** (max_depth - depth + 1))

            count = 1  # Count current directory

            try:
                for item in current_path.iterdir():
                    if item.is_dir() and is_accessible_directory(item):
                        count += count_directories(item, depth + 1)
            except (PermissionError, OSError):
                # If we can't access, estimate based on typical directory structure
                return max(1, int(10 * (1.5 ** (max_depth - depth))))

            return count

        total = count_directories(path, 0)
        return max(total, 1)  # At least 1 directory

    except Exception:
        # Fallback estimation
        return 1000

File: src/main/core/scanner.ts
/**
 * Directory scanning engine.
 *
 * @description Ported from Python `scanner.py`. Scans directories,
 *              classifies files, and aggregates statistics.
 *
 * @module main/core/scanner
 */

import { join } from 'path';
import { realpath } from 'fs/promises';
import type {
  DirectoryInfo,
  FileInfo,
  ScanOptions,
  ScanResult,
  ScanStatistics,
} from '@shared/types';
import { ContentClassifier } from './classifier';
import {
  isAccessibleDirectory,
  getDirectFiles,
  safeGetFileSize,
  getSubdirectories,
} from '@main/utils/fs';
import { ProgressReporter } from '@main/utils/progress';
import { cpus } from 'os';

/** Maximum concurrent async tasks for parallel scanning. */
const MAX_WORKERS = Math.max(1, cpus().length);

interface ScanDirectoryResult {
  dirInfo: DirectoryInfo;
  files: FileInfo[];
}

/**
 * Thrown when a scan is cancelled by the user.
 */
export class ScanCancelledError extends Error {
  constructor() {
    super('Scan cancelled by user');
    this.name = 'ScanCancelledError';
  }
}

/**
 * Orchestrates directory traversal, file classification, and result aggregation.
 */
export class DirectoryScanner {
  private static readonly SCAN_TIMEOUT_MS = 30 * 60 * 1000; // 30 minutes
  private static readonly MAX_FILES_LIMIT = 1_000_000;

  private readonly options: ScanOptions;
  private readonly classifier: ContentClassifier;
  private readonly errorDirectories: string[] = [];
  private readonly allFiles: FileInfo[] = [];
  private totalDirectories = 0;
  private totalFiles = 0;
  private totalSize = 0;
  private abortController: AbortController | null = null;
  private scanTimeoutId: NodeJS.Timeout | null = null;

  constructor(options: ScanOptions) {
    this.options = options;
    this.classifier = new ContentClassifier();
  }

  /**
   * Perform a complete scan and return a populated ScanResult.
   *
   * @param useParallel - Whether to process directories concurrently.
   * @param onProgress - Optional callback invoked with progress updates.
   * @returns ScanResult with sorted directories and statistics.
   * @throws {ScanCancelledError} If the scan is aborted.
   */
  async scan(
    useParallel = true,
    onProgress?: (current: number, total: number) => void
  ): Promise<ScanResult> {
    this.abortController = new AbortController();
    const signal = this.abortController.signal;
    this.scanTimeoutId = setTimeout(() => {
      this.abortController?.abort();
    }, DirectoryScanner.SCAN_TIMEOUT_MS);

    const startTime = Date.now();

    try {
      const allDirPaths = await this.collectAllDirectories(this.options.targetPath, signal);
      this.totalDirectories = allDirPaths.length;

      const reporter = new ProgressReporter(allDirPaths.length, 'Scanning', onProgress ?? null);

      let results: DirectoryInfo[];

      if (useParallel && allDirPaths.length > 1) {
        results = await this.scanParallel(allDirPaths, reporter, signal);
      } else {
        results = await this.scanSequential(allDirPaths, reporter, signal);
      }

      reporter.finish();

      results.sort((a, b) => b.sizeBytes - a.sizeBytes);
      results = results.slice(0, this.options.topCount);

      const duration = (Date.now() - startTime) / 1000;
      const statistics = this.createScanStatistics(results, duration);

      return {
        directories: results,
        totalScanned: this.totalDirectories,
        errorCount: this.errorDirectories.length,
        accessErrors: [...this.errorDirectories],
        scanDuration: duration,
        scanOptions: this.options,
        statistics,
      };
    } catch (err) {
      if (err instanceof ScanCancelledError) {
        throw err;
      }
      throw new Error(`Scan failed: ${err instanceof Error ? err.message : String(err)}`);
    } finally {
      clearTimeout(this.scanTimeoutId);
      this.scanTimeoutId = null;
      this.abortController = null;
    }
  }

  /**
   * Cancel an in-flight scan.
   */
  cancel(): void {
    this.abortController?.abort();
  }

  /**
   * Collect every directory under the root iteratively (avoids stack overflow).
   */
  private async collectAllDirectories(
    rootPath: string,
    signal: AbortSignal,
    maxDepth = 50,
    maxDirectories = 1_000_000
  ): Promise<string[]> {
    const dirs: string[] = [rootPath];
    const stack: Array<{ path: string; depth: number }> = [{ path: rootPath, depth: 0 }];
    const visited = new Set<string>();

    while (stack.length > 0) {
      if (signal.aborted) {
        throw new ScanCancelledError();
      }

      if (dirs.length >= maxDirectories) {
        break;
      }

      const item = stack.pop();
      if (!item) break;
      const { path: current, depth } = item;
      if (depth >= maxDepth) {
        continue;
      }

      try {
        const subdirs = await getSubdirectoryPaths(current, this.options.includeHidden);
        for (const sub of subdirs) {
          const full = join(current, sub);
          try {
            const realFull = await realpath(full);
            if (visited.has(realFull)) {
              continue;
            }
            visited.add(realFull);
          } catch {
            // If realpath fails, fall back to the raw path and continue.
          }
          dirs.push(full);
          stack.push({ path: full, depth: depth + 1 });
        }
      } catch {
        this.errorDirectories.push(current);
      }
    }

    return dirs;
  }

  /**
   * Scan directories one at a time.
   */
  private async scanSequential(
    dirPaths: string[],
    reporter: ProgressReporter,
    signal: AbortSignal
  ): Promise<DirectoryInfo[]> {
    const results: DirectoryInfo[] = [];

    for (const dirPath of dirPaths) {
      if (signal.aborted) {
        throw new ScanCancelledError();
      }

      const { dirInfo } = await this.scanSingleDirectory(dirPath);
      if (dirInfo.sizeBytes >= this.options.minSizeBytes) {
        results.push(dirInfo);
      }
      reporter.update();
    }

    return results;
  }

  /**
   * Scan directories with limited concurrency.
   *
   * @description Uses an async task pool instead of Worker Threads because
   *              filesystem scanning is I/O-bound. Node.js async I/O is
   *              already non-blocking; Worker Threads add overhead without
   *              benefit for this workload.
   */
  private async scanParallel(
    dirPaths: string[],
    reporter: ProgressReporter,
    signal: AbortSignal
  ): Promise<DirectoryInfo[]> {
    const results: DirectoryInfo[] = [];
    const queue = [...dirPaths];
    const inFlight = new Set<Promise<void>>();

    const processOne = async (dirPath: string): Promise<void> => {
      if (signal.aborted) {
        throw new ScanCancelledError();
      }
      const { dirInfo } = await this.scanSingleDirectory(dirPath);
      if (dirInfo.sizeBytes >= this.options.minSizeBytes) {
        results.push(dirInfo);
      }
      reporter.update();
    };

    while (queue.length > 0 || inFlight.size > 0) {
      if (signal.aborted) {
        throw new ScanCancelledError();
      }

      while (inFlight.size < MAX_WORKERS && queue.length > 0) {
        const dirPath = queue.shift();
        if (!dirPath) break;
        const task = processOne(dirPath).finally(() => {
          inFlight.delete(task);
        });
        inFlight.add(task);
      }

      if (inFlight.size >= MAX_WORKERS || queue.length === 0) {
        await Promise.race(inFlight);
      }
    }

    return results;
  }

  /**
   * Scan a single directory: sum direct file sizes, classify files, apply filters.
   */
  private async scanSingleDirectory(dirPath: string): Promise<ScanDirectoryResult> {
    const scanTime = Date.now();

    const isAccessible = await isAccessibleDirectory(dirPath);
    if (!isAccessible) {
      this.errorDirectories.push(dirPath);
      return {
        dirInfo: {
          path: dirPath,
          sizeBytes: 0,
          fileCount: 0,
          lastScanned: scanTime,
          errorMessage: 'Permission denied or directory inaccessible',
          files: [],
          categoryBreakdown: {},
          dominantCategory: null,
        },
        files: [],
      };
    }

    let totalSize = 0;
    let fileCount = 0;
    const directoryFiles: FileInfo[] = [];

    try {
      for await (const fileName of getDirectFiles(dirPath)) {
        const filePath = join(dirPath, fileName);
        const extension = getExtension(fileName);

        if (this.options.extensionFilter !== null && this.options.extensionFilter.length > 0) {
          if (!this.options.extensionFilter.includes(extension)) {
            continue;
          }
        }

        const fileSize = await safeGetFileSize(filePath);
        totalSize += fileSize;
        fileCount += 1;

        const fileInfo = this.classifier.classifyFileWithInfo(filePath, fileSize);
        directoryFiles.push(fileInfo);
        if (this.allFiles.length < DirectoryScanner.MAX_FILES_LIMIT) {
          this.allFiles.push(fileInfo);
        }
        this.totalFiles += 1;
        this.totalSize += fileSize;
      }
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      this.errorDirectories.push(dirPath);
      return {
        dirInfo: {
          path: dirPath,
          sizeBytes: 0,
          fileCount: 0,
          lastScanned: scanTime,
          errorMessage: msg,
          files: [],
          categoryBreakdown: {},
          dominantCategory: null,
        },
        files: [],
      };
    }

    const categoryBreakdown: Record<string, number> = {};
    let dominantCategory: string | null = null;
    let dominantSize = 0;

    for (const f of directoryFiles) {
      categoryBreakdown[f.category] = (categoryBreakdown[f.category] ?? 0) + f.sizeBytes;
      if (categoryBreakdown[f.category] > dominantSize) {
        dominantSize = categoryBreakdown[f.category];
        dominantCategory = f.category;
      }
    }

    return {
      dirInfo: {
        path: dirPath,
        sizeBytes: totalSize,
        fileCount,
        lastScanned: scanTime,
        errorMessage: null,
        files: directoryFiles,
        categoryBreakdown,
        dominantCategory,
      },
      files: directoryFiles,
    };
  }

  /**
   * Build aggregate statistics from the collected file data.
   */
  private createScanStatistics(_results: DirectoryInfo[], duration: number): ScanStatistics {
    return {
      totalDirectories: this.totalDirectories,
      totalFiles: this.totalFiles,
      totalSizeBytes: this.totalSize,
      scanDuration: duration,
      categoryBreakdown: this.classifier.getCategoryStatistics(this.allFiles),
      fileCountByCategory: this.classifier.getFileCountByCategory(this.allFiles),
    };
  }
}

/**
 * Helper: get normalized extension from a file name.
 */
function getExtension(fileName: string): string {
  const idx = fileName.lastIndexOf('.');
  return idx >= 0 ? fileName.slice(idx).toLowerCase() : '';
}

/**
 * Helper: list subdirectory names in a directory.
 */
async function getSubdirectoryPaths(parent: string, includeHidden: boolean): Promise<string[]> {
  const names: string[] = [];
  for await (const name of getSubdirectories(parent, includeHidden)) {
    names.push(name);
  }
  return names;
}
